### ISSUE TITLE
LDpred 1.0.0 using much more memory than previous versions

### ISSUE THREAD

**Issue by daniel-hui** [bvilhjal/ldpred#46]
Hi Bjarni, thank you for continually developing LDpred. We were using the version of LDpred from July 31, 2018 (v0.9), and the step to make the ldfile (now 'gibbs', was 'ldpred') was using about 250GB RAM with an LD radius of 2300 and about 7M SNPs, 504 individuals (1KG EUR). We've now downloaded a new version on Feb 1 this year (v.1.0.0), however this step now uses more than 500GB RAM (LD radius 1500, 4.5M SNPs, still same individuals), and we are unable to run it on our largest server. We sort of had the impression the updated versions were mostly for bug fixes and usability, but if this is not the case we are wondering if we should rerun our old analyses using the new version for consistency, or instead just use the old version for our new analyses. Specifically, when/which version of LDpred was updated to make this change in memory usage.

Going off the first issue, we are wondering if you could elaborate on your recommendation of reducing the analysis to HapMap3 SNPs (https://github.com/bvilhjal/ldpred/issues/36), similar to LDSC. Do you have thoughts on other filters such as some combination of those that are imputed well (eg. INFO > .3, or INFO > .9 or something), have some significant p-value (eg. p < .1), maybe those remaining after some light LD-based pruning, or those that are not strand ambiguous (as well as the MAF > .01 filter by default)? We are mostly interested in just running LDpred using default settings, since in the previous version of LDpred we would just use all SNPs that are available from a trait's GWAS but with the new version of LDpred this is not feasible -- we would just like to be certain of what preprocessing to apply since we will run LDpred across many different scenarios and would not like to have to redo things and benchmark different parameters outside of what is default/recommended.

To rephrase, there are three main questions:
1. Does LDpred from Feb 1 2019 (v1.0.0) have substantial analytical differences between LDpred from July 31 2018 (v0.9)?
2. Can you further elaborate on SNP filters to reduce the initial dataset, and how much they will impact results?
3. Is it prudent to just use an older version of LDpred (that does not use as much memory) with all SNPs in lieu of LDpred v.1.0.0 with a reduced set of SNPs -- specifically which older version would be recommended?

Thanks -- Daniel

**Comment by bvilhjal** [comment]
Hi Daniel, thanks a lot for this, and sorry about the slow reply.

I will look into the memory issue.  

Regarding the restricting to HapMap3 SNPs, there are two reasons for that.  First, is that LDpred is currently only well suited for analyzing about 200K-2M SNPs.  Hence, I suggest restricting to relatively common SNPs (because LD information for rare SNPs is poor), say MAF of at least 0.01.  I also suggest restricting to SNPs with low genotyping error rates (high info scores).  HapMap3 SNPs are likely to be suffice all of these criteria in most cases.

Regarding your questions.
1. The theory and the method is the same, but there are lots of bug fixes, ease-of-use improvements, and a shift to python 3. 
2. I believe I addressed that above.
3. Yes, you can do that.

Cheers,
Bjarni

**Comment by bvilhjal** [comment]
Hi, 

I added a flag (--only-hm3) in the coord step that enables user to easily restrict to HapMap3 SNPs.  I also proposed guideliens for the LD reference panel in the new Q and A wiki https://github.com/bvilhjal/ldpred/wiki which should ensure reasonable memory usage.  I will therefore close this issue. 

Best,
Bjarni

**Comment by wavefancy** [comment]
Hi Bjarni,

I found the high memory usage is because of the python3 pickle protocol used. Currently, the protocol is 2. ```pickle.dump(ld_dict, f, protocol=2) in ld.py file``` I changed this to ```pickle.dump(ld_dict, f, protocol=-1)```, which can save a huge amount of memory, reduce to more than half.  However, I would suggest using hickle. When I use [hickle](
https://github.com/telegraphic/hickle), the memory will not increase after we got the LD table. (protocol -1 will use double memory after having LD table ready).

Best regards
Wallace
